File: Cocoa Programming/Cocoa Pr L63 ObjC/Cocoa Pr L63 ObjC/DrawingView.m
//
//  DrawingView.m
//  Cocoa Pr L63 ObjC
//
//  Created by Michele Galvagno on 10/06/22.
//

#import "DrawingView.h"
#import "CircleView.h"

@implementation DrawingView

- (void)addCircle:(CircleView *)circle {
    [self addSubview:circle];
    [self.undoManager registerUndoWithTarget:self selector:@selector(removeCircle:) object:circle];
    [self.undoManager setActionName:@"Add Circle"];
}

- (void)removeCircle:(CircleView *)circle {
    [circle removeFromSuperview];
    [self.undoManager registerUndoWithTarget:self selector:@selector(addCircle:) object:circle];
    [self.undoManager setActionName:@"Remove Circle"];
}

- (void)mouseDown:(NSEvent *)event {
    NSPoint point = event.locationInWindow;
    NSUInteger flags = [event modifierFlags] & NSEventModifierFlagDeviceIndependentFlagsMask;
    
    if (flags == NSEventModifierFlagOption) {
        NSView *view = [self hitTest:point];
        [self removeCircle:(CircleView *)view];
    } else {
        CircleView *circle = [[CircleView alloc] initWithFrame:NSMakeRect(point.x-15, point.y-15, 30, 30)];
        [self addCircle:circle];
    }
}

- (BOOL)wantsUpdateLayer {
    return YES;
}

- (void)updateLayer {
    self.layer.backgroundColor = NSColor.blackColor.CGColor;
}
    
@end

File: Cocoa Programming/Cocoa Pr L63 NSUndoManager/Cocoa Pr L63 NSUndoManager/DrawingView.swift
//
//  DrawingView.swift
//  Cocoa Pr L63 NSUndoManager
//
//  Created by Michele Galvagno on 09/06/22.
//

import Cocoa

class DrawingView: NSView {

    override func mouseDown(with event: NSEvent) {
        let point = convert(event.locationInWindow, from: nil)
        if event.modifierFlags.contains(.option) {
            // Remove circle
            if let view = hitTest(point), view is CircleView {
                removeCircle(view)
                undoManager?.setActionName("Remove Circle")
            }
        } else {
            // Add circle
            let circle = CircleView(frame: NSMakeRect(point.x-15, point.y-15, 30, 30))
            addCircle(circle)
            undoManager?.setActionName("Add Circle")
        }
    }
    
    @objc func addCircle(_ circle: NSView) {
        addSubview(circle)
        undoManager?.registerUndo(withTarget: self, selector: #selector(self.removeCircle(_:)), object: circle)
    }
    
    @objc func removeCircle(_ circle: NSView) {
        circle.removeFromSuperview()
//        undoManager?.prepare(withInvocationTarget: self).addCircle(circle)
        undoManager?.registerUndo(withTarget: self, selector: #selector(self.addCircle(_:)), object: circle)
    }
    
    override var wantsUpdateLayer: Bool {
        return true
    }
    
    override func updateLayer() {
        layer?.backgroundColor = NSColor.black.cgColor
    }
    
}
